doc/man/gnunet-testbed.1
.TH GNUNET-TESTBED "1" "9 Dec 2003" "GNUnet"
.SH NAME
gnunet-testbed \- interactive environment for GNUnet simulations
.SH SYNOPSIS
.B gnunet\-testbed
[\fIOPTIONS\fR]
.SH DESCRIPTION
.PP
gnunet-testbed can be used to control GNUnet peers to perform testing and benchmarking of the system.
Note that gnunet-testbed is at this point still a very new tool and not functional.
gnunet-testbed is fully scriptable and uses bash as a scripting language.  All features of bash
are supported, including the execution of arbitrary shell commands.  gnunet-testbed extends bash
with additional, testbed specific commands.
.TP
\fB\-h\fR, \fB\-\-help\fR
print help page
.TP
\fB\-c \fIFILENAME\fR, \fB\-\-config=FILENAME\fR
load config file (defaults: ~/.gnunet/gnunet.conf)
.TP
\fB\-v\fR, \fB\-\-version\fR
print the version number
.TP
\fB\-L \fILOGLEVEL\fR, \fB\-\-loglevel=LOGLEVEL\fR
change the loglevel. Possible values for LOGLEVEL are NOTHING, FATAL, ERROR, FAILURE, WARNING, MESSAGE, INFO, DEBUG, CRON and EVERYTHING.
.SH NOTES
gnunet-testbed is not yet fully implemented.  Peers that participate in a testbed must load the
testbed module.
.SH FILES
.TP
~/.gnunet/gnunet.conf
GNUnet configuration file
.SH "REPORTING BUGS"
Report bugs by using mantis <https://gnunet.org/bugs/> or by sending electronic mail to <gnunet-developers@gnu.org>
.SH "SEE ALSO"
\fBgnunet.conf\fP(5), \fBgnunetd\fP(1)


ChangeLog
Mon Nov 16 09:25:47 CET 2009
	Allow passing of command-line arguments using GNUNET_ARGS
	environment variable (see Mantis #1467).

Mon Apr 27 22:18:19 CEST 2009
	Monitor low-traffic connections over a longer period
	of time before terminating them.

Fri Feb 27 22:05:50 MST 2009
	Releasing 0.8.0c.

Sun Feb  1 19:51:40 MST 2009
	Fixed Mantis #1429 (struct padding could cause sizeof
	to be different on AMD64, using gcc-ism "packed" to
	avoid this problem).

Sun Feb  1 00:37:16 MST 2009
	Fixed problem in DHT client protocol that could cause
	PUT requests from clients to be (frequently) lost.

Sun Nov  2 16:06:47 MST 2008
	Fixed problem with MySQL disconnects in one module
	impacting MySQL correctness in another module (caused
	crashes).

Sat Aug 16 00:18:34 CEST 2008
	Improved keyword extraction
	Fixed expiration time of keywords
	Releasing 0.8.0b.

Sun Jul 20 15:27:11 MDT 2008
	Releasing 0.8.0a.

Thu Jul 10 00:59:31 MDT 2008
	Fixed bug where datastore resizing would not 
	result in bloomfilter update during gnunet-update.

Sat Jun 28 23:43:00 MDT 2008
	Fixed bug preventing namespace advertisements from
	being processed.

Sat Jun 14 00:54:39 MDT 2008
	Releasing 0.8.0.

Fri Jun 13 23:51:00 MDT 2008
	GNUnet now compiles on OpenBSD.

Wed Jun 11 22:17:09 MDT 2008
	Made state of gnunet-auto-share persistent. 
	Added support for multiple directories.  Improved
	FSUI/ECRS APIs to reduce the number of threads needed
	for probing.  Increased number of concurrent probes
	allowed.

Sat Jun  7 17:54:49 MDT 2008
	Simplified namespaces a lot.  Shorter timeouts for
	trying to connect to gnunetd to prevent interactive
	apps from blocking too long when gnunetd is not
	running.

Sun Jun  1 21:11:54 MDT 2008
	Releasing 0.8.0pre1.

Sun Jun  1 13:41:27 MDT 2008
	Split of libgnunetpseudonym from libgnunetnamespace
	(with major API extension and changes).

Tue May  6 04:10:02 MDT 2008
	Added simple sanity check for [MODULES] values.
	Added "-s" and "-q" options to gnunet-peer-info.

Thu Apr 24 00:05:17 MDT 2008
	Simplified FSUI stop/abort/pause/resume API calls
	(no need to pass FSUI context anymore).

Mon Apr 21 21:05:20 MDT 2008
	Adding buffered IO for FSUI's serialize and
	deserialize code.

Sun Apr 20 20:01:20 MDT 2008
	Fixed bug in HTTP transport causing lonely messages
	(responses to GET) not to be delivered.

Mon Mar 24 21:15:36 MDT 2008
	Fixed various problems with downloading locally
	indexed large files (downloading large files from
	other peers should not have been a problem).
	
Sun Mar  2 15:33:41 MST 2008
	Added option to allow user to disable IPv6 support.
	Releasing 0.8.0pre0.

Sat Mar  1 21:14:02 MST 2008
	Fixed some dstore performance issues.
	Made transports alternate between trying IPv4 and IPv6
	if both are available (since one of them maybe
	misconfigured).

Mon Feb 25 23:25:48 MST 2008
	Cleanup of the DNS code: removed GNUNET_IPvXAddress
	and GNUNET_PRIP; centralized all DNS lookup code in 
	util/network/dns.c; improved IPv6 support for 
        network-client code.

Mon Feb 25 00:01:27 MST 2008
	Added asynchronous search and download methods for 
	ECRS library.  FSUI now can do with only one thread
	per search or download (until now, we had two 
	threads per search / download).

Tue Feb 19 20:35:28 MST 2008
	Updated database schemata to support O(1) operations
	even if there are N files under the same keywords.
	This breaks database backwards compatibility (and
	no migration code is provided).

Mon Feb 18 19:47:37 MST 2008
	Added new options -u and -s to gnunet-insert.

Sat Feb 16 21:37:33 MST 2008
	Implemented gnunet-auto-share for automatic
	sharing of directories.

Thu Feb 14 01:02:32 MST 2008
	Made HTTP transport work over IPv6 as well.

Wed Feb 13 22:59:07 MST 2008
	Modified mySQL and SQLite datastores to return diverse
	result sets during (partial) hash-based get iterations
	(should result in more diverse keyword result sets).

Tue Feb 12 23:54:34 MST 2008
	First shot at integrated hostlist HTTP server based
	on libmicrohttpd.

Sat Feb  9 22:06:23 MST 2008
	Minor changes to the ECRS/FSUI APIs for searches
	(no more timeouts and/or max-results).  Clients who
	need these features could easily implement them 
	directly (and they are not really useful to begin
	with).  
	Also, the IPv4 and IPv6 protocols are being integrated
	so that we only listen to one port (running both IPv4
	and IPv6 over the same port).  This break transport
	compatibility, but given that we're breaking protocol
	compatibility for file-sharing already, this may 
	actually be a bit of a good thing (since it will now
	be possible to tell if peers running the new GAP code
	are connected to other peers running the new GAP code).
	
Sun Feb  3 13:17:09 MST 2008
	Dramatic changes to the GAP implementation (breaking
	protocol compatibility).  Essentially, we can save
	a few bytes in each reply.  More importantly, the new
	code allows the searching client to specify a set of
	replies that are not desired (hopefully helping to
	dramatically increase the diversity of search replies
	obtained over time).  Note that the actual encoding
	and databases are not affected (just P2P protocol).
	The update is not complete yet, but should compile.

Tue Jan  8 20:07:20 MST 2008
	Added option for testing ("make check") to use weak(er)
	PRNG for key generation (thanks to Werner Koch for 
	pointing out how to do it).

Thu Dec 22 20:10:37 MST 2007
	Releasing GNUnet 0.7.3.

Sun Dec  9 14:34:32 MST 2007
	Implemented MySQL version of the dstore-module.  This
	means that sqLite is now again truly optional.

Sat Dec  8 15:15:53 MST 2007
	Integrated F2F topology into main topology module 
	(options are used to choose between F2F-only (old F2F), 
	 minimum number of friend connections (NEW) or entirely
	 arbitrary connection set (default)).

Thu Dec  6 20:51:17 MST 2007
	The SMTP transport is working again.

Thu Nov 22 20:49:16 MST 2007
	United all libgnunetutil_*.so libraries into one big
	libgnunetutil.so library (eliminates issues for binary

src/applications/testbed/gnunet-testbed.c
/*
     This file is part of GNUnet.
     (C) 2003, 2004, 2005 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file applications/testbed/gnunet-testbed.c
 * @brief A Testbed tool for performing distributed experiments
 * @author Ronaldo Alves Ferreira
 * @author Christian Grothoff
 * @author Murali Krishan Ramanathan
 *
 * Todo:
 * - test secure GNUNET_RSA_sign-on (and test testbed setup script!)
 * - allow removing of peers (explicitly AND when peer shuts down!)
 *   Problem: what happens to the peer-IDs in that case?
 * - general problem: the way we use tcpio means that any rouge
 *   testbed-gnunetd can stall gnunet-testbed indefinitely!
 * - security: limit "exec" to certain processes
 */

#include "gnunet_util.h"
#include "platform.h"
#include "testbed.h"
#include "socket.h"
#include "commands.h"
#include <signal.h>

#ifndef sighandler_t
typedef void (*sighandler_t) (int);
#endif

#define TESTBED_VERSION  	"0.0.5"
#define HELPER                  "==HELPER=="


/* we may want to change SHELL and PORT into values
   obtained from the configuration... */

#define SHELL  (NULL == getenv("BASH") ? "/bin/bash" : getenv("BASH"))

#define PORT 4242

/* TB_ALIASES should probably be forced to live somewhere
   under ~/.gnunet */
#define TB_ALIASES "/tmp/gnunet-testbed-aliasdefinitions"


/**
 * Name of gnunet-testbed binary.
 */
static char *testbedArg0;


/* ****************** scriptability *************** */

#ifndef MINGW                   /* FIXME MINGW */



/**
 * Parse the options, set the timeout.
 *
 * @param argc the number of options
 * @param argv the option list (including keywords)
 * @return GNUNET_OK on error, GNUNET_SYSERR if we should exit
 */
static int
helperParseOptions (int argc, char *argv[])
{
  int c, option_index;

  GNUNET_free_non_null (setConfigurationString ("GNUNETD", "LOGFILE", NULL));
  while (1)
    {
      static struct GNoption long_options[] = {
        LONG_DEFAULT_OPTIONS,
        {0, 0, 0, 0}
      };

      option_index = 0;
      c = GNgetopt_long (argc, argv, "vhdc:L:", long_options, &option_index);
      if (c == -1)
        break;                  /* No more flags to process */
      if (GNUNET_YES == parseDefaultOptions (c, GNoptarg))
        continue;
      switch (c)
        {
        case 'v':
          printf ("GNUnet v%s, gnunet-testbed v%s\n",
                  VERSION, TESTBED_VERSION);
          return GNUNET_SYSERR;
        case 'h':
          {
            static Help help[] = {
              HELP_CONFIG,
              HELP_HELP,
              HELP_LOGLEVEL,
              HELP_VERSION,
              HELP_END,
            };
            formatHelp ("gnunet-testbed ==HELPER== [OPTIONS] [COMMAND]",
                        _("Start GNUnet-testbed helper."), help);
            return GNUNET_SYSERR;
          }
        default:
          GNUNET_GE_LOG (ectx,
                         GNUNET_GE_ERROR | GNUNET_GE_IMMEDIATE |
                         GNUNET_GE_USER,
                         _("Use --help to get a list of options.\n"));
          return -1;
        }                       /* end of parsing commandline */
    }                           /* while (1) */
  setConfigurationStringList (&argv[GNoptind], argc - GNoptind);
  return GNUNET_OK;
}


/**
 * This is the main method of the helper-process that
 * is invoked from the bash-process.  Helper encapsulates
 * the command from the shell in a stream, sends it
 * over the socket to the main gnunet-testbed process,
 * retrieves there result and outputs he result back
 * to bash.
 */
static int
helper_main (int argc, char *argv[])
{
  int i;
  int retVal;
  unsigned int len;
  int res;
  char *buf;
  struct sockaddr_in soaddr;

  if (GNUNET_SYSERR == initUtil (argc, argv, &helperParseOptions))
    return -1;

  argc = getConfigurationStringList (&argv);

  if (argc == 0)
    {
      fprintf (stderr, " must have at least one argument!\n");
      return -1;
    }
  sock = SOCKET (PF_INET, SOCK_STREAM, 6);      /* 6: TCP */
  if (sock == -1)
    {
      LOG_STRERROR (LOG_FAILURE, "socket");
      return GNUNET_SYSERR;
    }
  soaddr.sin_family = AF_INET;
  soaddr.sin_addr.s_addr = htonl (INADDR_LOOPBACK);
  soaddr.sin_port = htons (PORT);
  res = CONNECT (sock, (struct sockaddr *) &soaddr, sizeof (soaddr));
  if ((res < 0) && (errno != EINPROGRESS) && (errno != EWOULDBLOCK))
    {
      GNUNET_GE_LOG (ectx,
                     GNUNET_GE_INFO | GNUNET_GE_REQUEST | GNUNET_GE_USER,
                     _("Cannot connect to LOOPBACK port %d: %s\n"), PORT,
                     STRERROR (errno));
      closefile (sock);
      sock = -1;
      return GNUNET_SYSERR;
    }

  /* write command to socket */
  socketSend (strlen (argv[0]), SOCKET_BEGIN_COMMAND, argv[0]);
  GNUNET_free (argv[0]);
  /* write args to socket */
  for (i = 1; i < argc; i++)
    {
      socketSend (strlen (argv[i]), SOCKET_ADD_ARGUMENT, argv[i]);
      GNUNET_free (argv[i]);
    }
  GNUNET_free (argv);
  socketSend (0, SOCKET_END_COMMAND, NULL);

  /* read result from socket, print to stderr,  obtain retVal */
  i = SOCKET_PRINTF;
  buf = NULL;
  while (i == SOCKET_PRINTF)
    {
      GNUNET_free_non_null (buf);
      buf = NULL;
      i = readSocket (&buf, &len);
      if (i == SOCKET_PRINTF)
        fprintf (stdout, "%.*s", (int) len, buf);
    }
  retVal = *(int *) buf;
  GNUNET_free (buf);
  closefile (sock);
  return retVal;
}

/**
 * A child (the shell) has quit.  So we exit,
 * too.
 */
static void
sigChildHandler (int signal, siginfo_t * info, void *extra)
{
  do_quit = GNUNET_YES;
}

/**
 * This is the "bash" process.  Execs bash.
 * @returns never
 */
static void
bash_main ()
{
  int i;
  FILE *aliases;
  char *configFile;
  char *argv[] = {
    NULL,                       /* replaced by SHELL */
    "--init-file",
    TB_ALIASES,
    "-i",
    NULL,
  };

  configFile = getConfigurationString ("FILES", "gnunet.conf");
  GNUNET_GE_ASSERT (ectx, configFile != NULL);
  argv[0] = SHELL;
  aliases = FOPEN (TB_ALIASES, "w+");
  fprintf (aliases, "export PS1=\"[GTB]%% \"\n");
  i = 0;
  while (commands[i].command != NULL)
    {
      if (0 == strcmp ("exit", commands[i].command))
        {
          fprintf (aliases,
                   "alias exit=\"%s ==HELPER== -c %s exit ; exit\"\n",
                   testbedArg0, configFile);
        }
      else
        {
          fprintf (aliases,
                   "alias %s=\"%s ==HELPER== -c %s %s\"\n",
                   commands[i].command,
                   testbedArg0, configFile, commands[i].command);
        }
      i++;
    }
  GNUNET_free (configFile);
  fclose (aliases);
  doneUtil ();
  execvp (SHELL, argv);
  fprintf (stderr,
           _("Could not execute `%s': %s\n"), SHELL, STRERROR (errno));
}

/**
 * Configuration...
 */
static struct GNUNET_IPv4NetworkSet *trustedNetworks_ = NULL;

/**
 * Is this IP labeled as trusted for CS connections?
 */
static int
isWhitelisted (GNUNET_IPv4Address ip)
{
  return checkIPListed (trustedNetworks_, ip);
}

/**
 * This is the main method of the server.  It reads
 * commands from the socket that are created by helper process.
 * It is the main process that also keeps the state
 * throughout the session.
 *
 * @param bash_pid the process ID of the child that is bash
 */
static int
server_main (pid_t bash_pid)
{
  int i, status, ssock;
  socklen_t lenOfIncomingAddr;
  int secs = 5;
  const int on = 1;
  struct sockaddr_in serverAddr, clientAddr;
  struct sigaction oldAct;
  struct sigaction newAct;
  sigset_t set;
  sigset_t oset;


  /* create the socket */
CREATE_SOCKET:
  while ((ssock = SOCKET (PF_INET, SOCK_STREAM, 0)) < 0)
    {
      LOG_STRERROR (LOG_ERROR, "socket");
      GNUNET_GE_LOG (ectx, GNUNET_GE_ERROR | GNUNET_GE_BULK | GNUNET_GE_USER,
                     _
                     ("No client service started. Trying again in 30 seconds.\n"));
      sleep (30);
    }

  serverAddr.sin_family = AF_INET;
  serverAddr.sin_addr.s_addr = htonl (INADDR_ANY);
  serverAddr.sin_port = htons (PORT);

  if (SETSOCKOPT (ssock, SOL_SOCKET, SO_REUSEADDR, &on, sizeof (on)) < 0)
    LOG_STRERROR (LOG_ERROR, "setsockopt");

  /* bind the socket */
  if (BIND (ssock, (struct sockaddr *) &serverAddr, sizeof (serverAddr)) < 0)
    {
      GNUNET_GE_LOG (ectx, GNUNET_GE_ERROR | GNUNET_GE_BULK | GNUNET_GE_USER,
                     _("Error (%s) binding the TCP listener to "
                       "port %d. No proxy service started.\nTrying "
                       "again in %d seconds...\n"), STRERROR (errno), PORT,
                     secs);
      sleep (secs);
      secs += 5;                /* slow progression... */
      closefile (ssock);
      goto CREATE_SOCKET;
    }

  do_quit = GNUNET_NO;
  /* signal handler is needed if the child did exit
     (e.g. with CTRL-D and not with the command 'exit') */
  newAct.sa_sigaction = &sigChildHandler;
  sigfillset (&newAct.sa_mask);
  newAct.sa_flags = SA_NOCLDSTOP | SA_SIGINFO | SA_RESTART;
  if (0 != sigaction (SIGCHLD, &newAct, &oldAct))
    DIE_STRERROR ("sigaction");
  sigemptyset (&set);
  sigaddset (&set, SIGCHLD);
  if (0 != sigprocmask (SIG_UNBLOCK, &set, &oset))
    DIE_STRERROR ("sigprocmask");
  LISTEN (ssock, 5);
  while ((do_quit == GNUNET_NO)
         && (0 == waitpid (bash_pid, &status, WNOHANG)))
    {
      unsigned int argc;
      char *command;
      char **args;
      char *buf;
      unsigned int len;
      fd_set rset;
      fd_set wset;
      fd_set eset;
      GNUNET_IPv4Address ipaddr;

      lenOfIncomingAddr = sizeof (clientAddr);
      /* accept is not interrupted by SIGCHLD,
         so we must do a select first; yuck. */
      FD_ZERO (&rset);
      FD_ZERO (&wset);
      FD_ZERO (&eset);
      FD_SET (ssock, &rset);
      sock = SELECT (ssock + 1, &rset, &wset, &eset, NULL);
      if (sock == -1)
        continue;
      sock = ACCEPT (ssock,
                     (struct sockaddr *) &clientAddr, &lenOfIncomingAddr);
      if (sock < 0)
        {
          LOG_STRERROR (LOG_ERROR, "accept");
          continue;
        }
      /* access control! */
      GNUNET_GE_ASSERT (ectx,
                        sizeof (struct in_addr) ==
                        sizeof (GNUNET_IPv4Address));
      memcpy (&ipaddr, &clientAddr.sin_addr, sizeof (struct in_addr));
      if (GNUNET_NO == isWhitelisted (ipaddr))
        {
          GNUNET_GE_LOG (ectx,
                         GNUNET_GE_WARNING | GNUNET_GE_BULK | GNUNET_GE_USER,
                         _
                         ("Rejected unauthorized connection from %u.%u.%u.%u.\n"),
                         PRIP (ntohl (*(int *) &clientAddr.sin_addr)));
          closefile (sock);
          continue;
        }


      /* read from socket, run reaction,
         return result value and stdout value;
         possibly set doExit to 1 to exit */
      buf = NULL;
      if (SOCKET_BEGIN_COMMAND != readSocket (&buf, &len))
        {
          fprintf (stderr,
                   _("Protocol violation on socket. " "Expected command.\n"));
          return -1;
        }
      command = GNUNET_malloc (len + 1);
      memcpy (command, buf, len);
      command[len] = '\0';
      argc = 0;
      args = NULL;
      GNUNET_free (buf);
      buf = NULL;
      while (SOCKET_ADD_ARGUMENT == readSocket (&buf, &len))
        {
          GNUNET_array_grow (args, argc, argc + 1);
          args[argc - 1] = GNUNET_malloc (len + 1);
          memcpy (args[argc - 1], buf, len);
          args[argc - 1][len] = '\0';
          GNUNET_free (buf);
          buf = NULL;
        }
      GNUNET_free_non_null (buf);
      i = 0;
      while (commands[i].command != NULL)
        {
          if (0 == strcmp (commands[i].command, command))
            {
              int ret;
              ret = commands[i].handler (argc, args);
              socketSend (sizeof (unsigned int), SOCKET_RETVAL, &ret);
              break;
            }
          i++;
        }
      for (i = 0; i < argc; i++)
        GNUNET_free (args[i]);
      GNUNET_array_grow (args, argc, 0);
      if (commands[i].command == NULL)
        {
          /* should never happen unless the user
             plays by hand with the aliases... */
          i = -1;
          PRINTF (_("Command `%s' not found!\n"), command);
          socketSend (sizeof (unsigned int), SOCKET_RETVAL, &i);
        }
      GNUNET_free (command);
      closefile (sock);
      sock = -1;
    }
  /* just to be certain, we could have
     left the main loop due to doExit... */
  waitpid (bash_pid, &status, WNOHANG);
  /* restore... */
  if (0 != sigaction (SIGCHLD, &oldAct, &newAct))
    LOG_STRERROR (LOG_WARNING, "sigaction");
  return status;
}

#endif

/* *************** command line options *********** */


/**
 * Parse the options, set the timeout.
 *
 * @param argc the number of options
 * @param argv the option list (including keywords)
 * @return GNUNET_OK on error, GNUNET_SYSERR if we should exit
 */
static int
parseOptions (int argc, char *argv[])
{
  int c, option_index;

  GNUNET_free_non_null (setConfigurationString ("GNUNETD", "LOGFILE", NULL));
  while (1)
    {
      static struct GNoption long_options[] = {
        LONG_DEFAULT_OPTIONS,
        {0, 0, 0, 0}
      };

      option_index = 0;
      c = GNgetopt_long (argc, argv, "vhdc:L:", long_options, &option_index);
      if (c == -1)
        break;                  /* No more flags to process */
      if (GNUNET_YES == parseDefaultOptions (c, GNoptarg))
        continue;
      switch (c)
        {
        case 'v':
          printf ("GNUnet v%s, gnunet-testbed v%s\n",
                  VERSION, TESTBED_VERSION);
          return GNUNET_SYSERR;

        case 'h':
          {
            static Help help[] = {
              HELP_CONFIG,
              HELP_HELP,
              HELP_LOGLEVEL,
              HELP_VERSION,
              HELP_END,
            };
            formatHelp ("gnunet-testbed [OPTIONS]",
                        _("Start GNUnet testbed controller."), help);
            return GNUNET_SYSERR;
          }
        default:
          GNUNET_GE_LOG (ectx,
                         GNUNET_GE_ERROR | GNUNET_GE_IMMEDIATE |
                         GNUNET_GE_USER,
                         _("Use --help to get a list of options.\n"));
          return -1;
        }                       /* end of parsing commandline */
    }                           /* while (1) */
  return GNUNET_OK;
}

/* **************** main **************** */

/**
 * Create a testbed environment with multiple peers to collect
 * GNUnet statistics.  Note that the same binary is used for
 * two distinct purposes.  One is the stateful master process
 * that maintains the testbed.  A second type of process is
 * created if HELPER is the first argument.  These processes
 * are created by the user from the SHELL and they merely serve
 * to communicate back the user-commands to the master-process.
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return 0 on success, -1 on error
 */
int
main (int argc, char *argv[])
{
#ifndef MINGW                   /* FIXME MINGW */
  pid_t pid;
  char *ch;

  testbedArg0 = expandFileName (argv[0]);
  /* first, evaluate if we are the special
     helper process.  If so, eat HELPER
     argument and run helper_main */
  if (argc > 1)
    {
      if (strcmp (argv[1], HELPER) == 0)
        {
          argv[1] = argv[0];
          return helper_main (argc - 1, &argv[1]);
        }
    }

  if (GNUNET_SYSERR == initUtil (argc, argv, &parseOptions))
    return -1;

  ch = getConfigurationString ("GNUNET-TESTBED", "TRUSTED");
  if (ch == NULL)
    {
      trustedNetworks_ = parseRoutes ("127.0.0.0/8;");  /* by default, trust localhost only */
    }
  else
    {
      trustedNetworks_ = parseRoutes (ch);
      if (trustedNetworks_ == NULL)
        errexit (_
                 ("Malformed entry in the configuration in section %s under %s: %s\n"),
                 "GNUNET-TESTBED", "TRUSTED", ch);
      GNUNET_free (ch);
    }

  /* we are the main testbed process.  Fork of
     a shell and start processing from the socket */

  pid = fork ();
  if (pid < 0)
    DIE_STRERROR ("fork");
  if (pid == 0)
    {
      GNUNET_free (trustedNetworks_);
      bash_main ();
      return 0;                 /* unreached */
    }
  else
    {
      int ret;
      /* run actual main loop */
      ret = server_main (pid);
      /* just to be certain */
      kill (pid, SIGHUP);
      /* proper shutdown... */
      doneUtil ();
      GNUNET_free (trustedNetworks_);
      UNLINK (TB_ALIASES);
      GNUNET_free (testbedArg0);
      return ret;
    }
#endif
}

/* end of gnunet-testbed.c */
